File: uniffi_bindgen/src/bindings/kotlin/templates/TopLevelFunctionTemplate.kt
{%- if func.is_async() %}
{%- match func.throws_type() -%}
{%- when Some with (throwable) %}
@Throws({{ throwable|error_type_name }}::class)
{%- else -%}
{%- endmatch %}

@Suppress("ASSIGNED_BUT_NEVER_ACCESSED_VARIABLE")
suspend fun {{ func.name()|fn_name }}({%- call kt::arg_list_decl(func) -%}){% match func.return_type() %}{% when Some with (return_type) %} : {{ return_type|type_or_iface_name }}{% when None %}{%- endmatch %} {
    return uniffiRustCallAsync(
        _UniFFILib.INSTANCE.{{ func.ffi_func().name() }}({% call kt::arg_list_lowered(func) %}),
        { future, continuation -> _UniFFILib.INSTANCE.{{ func.ffi_rust_future_poll(ci) }}(future, continuation) },
        { future, status -> _UniFFILib.INSTANCE.{{ func.ffi_rust_future_complete(ci) }}(future, status) },
        { future -> _UniFFILib.INSTANCE.{{ func.ffi_rust_future_free(ci) }}(future) },
        // lift function
        {%- match func.return_type() %}
        {%- when Some(return_type) %}
        { {{ return_type|lift_fn }}(it) },
        {%- when None %}
        { Unit },
        {% endmatch %}
        // Error FFI converter
        {%- match func.throws_type() %}
        {%- when Some(e) %}
        {{ e|error_type_name }}.ErrorHandler,
        {%- when None %}
        NullCallStatusErrorHandler,
        {%- endmatch %}
    )
}

{%- else %}
{%- match func.throws_type() -%}
{%- when Some with (throwable) %}
@Throws({{ throwable|error_type_name }}::class)
{%- else -%}
{%- endmatch -%}

{%- match func.return_type() -%}
{%- when Some with (return_type) %}

fun {{ func.name()|fn_name }}({%- call kt::arg_list_decl(func) -%}): {{ return_type|type_or_iface_name }} {
    return {{ return_type|lift_fn }}({% call kt::to_ffi_call(func) %})
}
{% when None %}

fun {{ func.name()|fn_name }}({% call kt::arg_list_decl(func) %}) =
    {% call kt::to_ffi_call(func) %}

{% endmatch %}
{%- endif %}

File: uniffi_bindgen/src/bindings/swift/templates/TopLevelFunctionTemplate.swift
{%- if func.is_async() %}

public func {{ func.name()|fn_name }}({%- call swift::arg_list_decl(func) -%}) async {% call swift::throws(func) %}{% match func.return_type() %}{% when Some with (return_type) %} -> {{ return_type|type_name }}{% when None %}{% endmatch %} {
    {%- if func|is_cancellable(config) %}
    return try await uniffiRustCallAsyncCancellable(
    {%- else %}
    return {% if func.throws() %}try {% endif %}await uniffiRustCallAsync(
    {%- endif %}
        rustFutureFunc: {
            {{ func.ffi_func().name() }}(
                {%- for arg in func.arguments() %}
                {{ arg|lower_fn }}({{ arg.name()|var_name }}){% if !loop.last %},{% endif %}
                {%- endfor %}
            )
        },
        pollFunc: {{ func.ffi_rust_future_poll(ci) }},
        {%- if func|is_cancellable(config) %}
        cancelFunc: {{ func.ffi_rust_future_cancel(ci) }},
        {%- endif %}
        completeFunc: { rustFuture in
            {%- match func.return_type() %}
            {%- when Some(return_type) %}
            let liftReturn = { try! {{ return_type|lift_fn }}($0) }
            {%- when None %}
            let liftReturn = { (_: ()) in () }
            {%- endmatch %}

            return liftReturn(
                {%- match func.throws_type() %}
                {%- when Some with (e) %}
                try uniffiRustCallWithError(
                    {{ e|ffi_converter_name }}.lift
                ) { callStatus in
                    {{ func.ffi_rust_future_complete(ci) }}(rustFuture, callStatus)
                }
                {%- else %}
                uniffiRustCall { callStatus in
                    {{ func.ffi_rust_future_complete(ci) }}(rustFuture, callStatus)
                }
                {%- endmatch %}
            )

        },
        freeFunc: {{ func.ffi_rust_future_free(ci) }}
    )
}

{% else %}

{%- match func.return_type() -%}
{%- when Some with (return_type) %}

public func {{ func.name()|fn_name }}({%- call swift::arg_list_decl(func) -%}) {% call swift::throws(func) %} -> {{ return_type|type_name }} {
    return {% call swift::try(func) %} {{ return_type|lift_fn }}(
        {% call swift::to_ffi_call(func) %}
    )
}

{%- when None %}

public func {{ func.name()|fn_name }}({% call swift::arg_list_decl(func) %}) {% call swift::throws(func) %} {
    {% call swift::to_ffi_call(func) %}
}

{% endmatch %}
{%- endif %}
